File: A_Star/Python/a_star.py
"""
    Filename: a_star.py
    Description: Plan path using A* Algorithm
    Author: Benxiaogu:https://github.com/Benxiaogu
    Date: 2024-08-05
"""
import heapq
import math
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation


class AStar:
    def __init__(self,grid,start,goal,board_size):
        self.grid = grid
        self.start = self.Node(start,0,0)
        self.goal = self.Node(goal,0,0)
        self.board_size = board_size

    class Node:
        def __init__(self, position, g, h, parent=None):
            self.position = position    # position of node
            self.g = g                  # distance from node to start
            self.h = h                  # heuristic value from node to goal
            self.parent = parent        # parent node

        def __eq__(self, other) -> bool:
            return self.position == other.position
        
        def __lt__(self, other):
            # Prioritise nodes based on heuristic value
            return self.g+self.h < other.g+other.h or (self.g+self.h==other.g+other.h and self.h<other.h)
        
    class Neighbor:
        def __init__(self,directions,cost):
            self.directions = directions
            self.cost = cost
    
    def get_neighbors(self, node):
        neighbors = []
        distances = []
        nexts = [self.Neighbor((-1, 0),1), self.Neighbor((0, 1),1), self.Neighbor((0, -1),1), self.Neighbor((1,0),1),
                self.Neighbor((-1,1),math.sqrt(2)), self.Neighbor((1,1),math.sqrt(2)),self.Neighbor((1, -1),math.sqrt(2)), self.Neighbor((-1,-1),math.sqrt(2))]
        for next in nexts:
            neighbor = (node[0] + next.directions[0], node[1] + next.directions[1])
            if self.board_size <= neighbor[0] < len(self.grid)-self.board_size and self.board_size <= neighbor[1] < len(self.grid[0])-self.board_size:
                if self.grid[neighbor[0]][neighbor[1]] == 0:
                    if next.directions[0] != 0 and next.directions[1] != 0:  # 对角线方向
                        if self.grid[node[0] + next.directions[0]][node[1]] == 0 and self.grid[node[0]][node[1] + next.directions[1]] == 0:
                            neighbors.append(neighbor)
                            distances.append(next.cost)
                    else:
                        neighbors.append(neighbor)
                        distances.append(next.cost)

        return neighbors,distances

    def plan(self):
        open = []
        self.searched = [] # Used to record nodes that are searched
        g_cost = {self.start.position:0}

        self.start.h = self.heuristic(self.start.position)
        heapq.heappush(open, self.start)

        while open:
            # Select the node closest to the start node
            current_node = heapq.heappop(open)
            if current_node.position in self.searched:
                continue
            self.searched.append(current_node.position)
            
            # Find the goal
            if current_node == self.goal:
                self.goal = current_node
                self.path = []
                while current_node:
                    self.path.append(current_node.position)
                    current_node = current_node.parent
                self.path = self.path[::-1]
                        
                return self.goal.g
            
            # Find the neighbors of the current node and determine in turn if they have already been closed
            neighbors, distances = self.get_neighbors(current_node.position)
            for neighbor,distance in zip(neighbors,distances):
                if neighbor in self.searched:
                    continue
                h = self.heuristic(neighbor)
                g = current_node.g+distance

                if neighbor not in g_cost or g < g_cost[neighbor]:
                    g_cost[neighbor] = g
                    neighbor_node = self.Node(neighbor,g,h,current_node)
                    heapq.heappush(open,neighbor_node)
                
        return -1

    # def heuristic(self, node):
    #     # Manhattan distance from current node to goal node
    #     return abs(node[0] - self.goal.position[0]) + abs(node[1] - self.goal.position[1])
    
    # def heuristic(self, node):
    #     # Chebyshev Distance
    #     D = 1
    #     D2 = math.sqrt(2)
    #     dx = abs(node[0] - self.goal.position[0])
    #     dy = abs(node[1] - self.goal.position[1])
    #     return D * (dx + dy) + (D2 - 2 * D) * min(dx, dy)
    
    def heuristic(self, node):
        # Euclidean Distance
        D = 1
        dx = abs(node[0] - self.goal.position[0])
        dy = abs(node[1] - self.goal.position[1])
        return D * math.sqrt(dx * dx + dy * dy)
    

    def visualize_grid(self, cost):
        fig, ax = plt.subplots()
        grid = np.array(self.grid)

        plt.imshow(grid, cmap='Greys', origin='upper')

        plt.title("AStar\n"+"Cost: "+str(cost))

        # Mark the start and goal point
        plt.scatter(self.start.position[1], self.start.position[0], c='green', marker='o', s=60)
        plt.scatter(self.goal.position[1], self.goal.position[0], c='blue', marker='o', s=60)

        # # Mark locations which has been visited
        # for node in self.searched:
        #     plt.gca().add_patch(plt.Rectangle((node[1]-0.5, node[0]-0.5), 1, 1, fill=True, color='gray', alpha=0.5))
            
        # # Mark path
        # if self.path:
        #     path_x, path_y = zip(*self.path)
        #     plt.plot(path_y, path_x, c='red', linewidth=2)

        
        visited_patches = []
        path_line, = ax.plot([], [], c='red', linewidth=2)
        # for order in range(len(self.searched)):
        #     node = self.searched[order]
        #     plt.Rectangle((node[1]-0.5, node[0]-0.5), 1, 1, fill=True, color='gray', alpha=0.5)

        def update(frame):
            if frame < len(self.searched):
                node = self.searched[frame]
                patch = plt.Rectangle((node[1]-0.5, node[0]-0.5), 1, 1, fill=True, color='gray', alpha=0.5)
                visited_patches.append(patch)
                ax.add_patch(patch)
            elif self.path:
                path_x, path_y = zip(*self.path)
                path_line.set_data(path_y, path_x)

            return visited_patches + [path_line]*10
    
        ani = animation.FuncAnimation(fig, update, frames=len(self.searched) + 10, interval=50, repeat=False)
        ani.save("map_animate.gif",writer='pillow')
        # ani.save("map_animate.mp4",writer='ffmpeg')

        plt.savefig("map_animate.png",bbox_inches='tight')
        plt.show()
        

File: A_Star/CPP/src/aStar.cpp
/**
 * @file aStar.cpp
 * @brief Plan path using A* Algorithm
 * @author Benxiaogu:https://github.com/Benxiaogu
 * @date 2024-08-05
**/

#include "aStar.hpp"
#include <iostream>
#include <queue>
#include <set>
#include <map>
#include <algorithm>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

AStar::AStar(vector<vector<int>> grid, pair<int, int> start, pair<int, int> goal, int radius, int board_size)
            :grid(grid),start(start),goal(goal),radius(radius),board_size(board_size)
{
}

std::tuple<vector<pair<int, int>>, vector<double>> AStar::get_neighbors(pair<int, int> node) 
{
    /*
        Search the neighbors of node
        return neighbors and corresponding distances
    
    */
    vector<pair<int, int>> neighbors;
    vector<double> distances;
    vector<Next> nexts = {Next({-1, 0}, 1), Next({-1, 1}, sqrt(2)), Next({0, 1}, 1), Next({1, 1}, sqrt(2)),
                          Next({1, 0}, 1), Next({1, -1}, sqrt(2)), Next({0, -1}, 1), Next({-1, -1}, sqrt(2))};
    for (auto& next : nexts) 
    {
        pair<int, int> neighbor = {node.first + next.directions.first, node.second + next.directions.second};
        if (0 <= neighbor.first && neighbor.first < grid.size() && 0 <= neighbor.second && neighbor.second < grid[0].size()) {
            if (grid[neighbor.first][neighbor.second] == 0) {
                if (next.directions.first != 0 && next.directions.second != 0) {
                    if (grid[node.first + next.directions.first][node.second] == 0 && grid[node.first][node.second + next.directions.second] == 0) {
                        neighbors.push_back(neighbor);
                        distances.push_back(next.cost);
                    }
                } 
                else {
                    neighbors.push_back(neighbor);
                    distances.push_back(next.cost);
                }
            }
        }
    }
    return make_tuple(neighbors,distances);
}

double AStar::plan()
{
    priority_queue<pair<double, pair<int, int>>, vector<pair<double, pair<int, int>>>, greater<>> open; // Priority queue is used to  prioritise of the nodes in priority queue based on the total cost of path exploration
    map<pair<int, int>, double> g_costs;    // G(n): the cost from current node to target node
    g_costs[start] = 0;
    open.push({heuristic(start),start});
    map<pair<int, int>, pair<int, int>> parents;

    while(!open.empty())
    {
        auto [current_f, current_node] = open.top();
        open.pop();

        if(closed.find(current_node) != closed.end()) continue; // Skip it if current node has been in closed
        closed.insert(current_node);

        if(current_node == goal){   // If it reaches the goal then construct the path by backtracking
            path.clear();
            while (current_node != start) 
            {
                path.push_back(current_node);
                current_node = parents[current_node];
            }
            path.push_back(start);
            reverse(path.begin(), path.end());
            
            return g_costs[goal];
        }

        visited.push_back(current_node);
        vector<pair<int,int>> neighbors;
        vector<double> distances;
        tie(neighbors,distances) = get_neighbors(current_node);
        for (size_t i=0; i<neighbors.size();++i) 
        {
            pair<int, int> neighbor = neighbors[i];
            if(closed.find(neighbor)!=closed.end()) continue;   //Skip it if the neighbor has been in closed

            double distance = distances[i];
            double g_cost = g_costs[current_node] + distance;
            double h_cost = heuristic(neighbor);
            double f_cost = g_cost + h_cost;
            if(g_costs.find(neighbor) == g_costs.end() || g_cost<g_costs[neighbor]){
                // g_costs = open + closed, then open = g_costs - closed;
                // check whether neighbor is in g_costs without closed equal to check whether neighbor is in open, 
                // if neighbor has been in open and new g_cost of neighbor is smaller then update it
                g_costs[neighbor] = g_cost;
                parents[neighbor] = current_node;
                open.push({f_cost,neighbor});
            }
        }
    }

    return -1;
}

double AStar::heuristic(pair<int,int> node)
{
    /*
        Compute estimated cost from "node" to target node
    */
    double dx = abs(node.first - goal.first);
    double dy = abs(node.second - goal.second);

    // Manhattan distance
    // return dx+dy;

    // Euclidean Distance
    return sqrt(dx*dx+dy*dy);
}

void AStar::visualize_grid(double cost) {
    int rows = grid.size();
    int cols = grid[0].size();
    
    // Set the size of grid
    int cell_size = 20;
    int img_width = cols * cell_size;
    int img_height = rows * cell_size;

    // Create video file
    // AVI Pattern
    // VideoWriter video("dijkstra_animation.avi", VideoWriter::fourcc('M', 'J', 'P', 'G'), 100, Size(img_width, img_height));
    // MP4 Pattern
    VideoWriter video("AStar_animation.mp4", VideoWriter::fourcc('m', 'p', '4', 'v'), 100, Size(img_width, img_height));
    
    // Create image for each frame
    Mat img(img_height, img_width, CV_8UC3, Scalar(255, 255, 255));
    // Title and Cost
    std::string title = "AStar";
    putText(img, title, Point(550, 40), FONT_HERSHEY_SIMPLEX, 0.8, Scalar(0, 0, 0), 2);
    std::string cost_text = "Cost: " + std::to_string(cost);
    putText(img, cost_text, Point(500, 80), FONT_HERSHEY_SIMPLEX, 0.8, Scalar(0, 0, 0), 2);

    // Draw Obstacles
    for (int i = 0; i < rows; ++i) 
    {
        for (int j = 0; j < cols; ++j) 
        {
            if (grid[i][j] == 1) {
                rectangle(img, Point(j * cell_size, i * cell_size),
                          Point((j + 1) * cell_size, (i + 1) * cell_size),
                          Scalar(0, 0, 0), FILLED);
            }
        }
    }

    // Draw start and goal point
    rectangle(img, Point(start.second * cell_size, start.first * cell_size),
            Point((start.second + 1) * cell_size, (start.first + 1) * cell_size),
            Scalar(0, 255, 0), FILLED);

    rectangle(img, Point(goal.second * cell_size, goal.first * cell_size),
            Point((goal.second + 1) * cell_size, (goal.first + 1) * cell_size),
            Scalar(0, 255, 255), FILLED);

    // Draw visited areas
    for (size_t i=1;i<visited.size()-1;++i) 
    {
        const auto& node = visited[i];
        rectangle(img, Point(node.second * cell_size, node.first * cell_size),
        Point((node.second + 1) * cell_size, (node.first + 1) * cell_size),
        Scalar(128, 128, 128), FILLED);
        
        video.write(img);
    }

    // Draw path
    for (size_t i = 1; i < path.size(); ++i) 
    {
        Point p1(path[i - 1].second * cell_size + cell_size / 2,
                path[i - 1].first * cell_size + cell_size / 2);
        Point p2(path[i].second * cell_size + cell_size / 2,
                path[i].first * cell_size + cell_size / 2);
        line(img, p1, p2, Scalar(0, 0, 255), 2);

        video.write(img);
    }

    video.release();
    imwrite("AStar_result.png", img);
    imshow("AStar", img);
    waitKey(0);
    destroyAllWindows();
}